Documentation/concepts/ipam/eni.rst

```diff
@@ -0,0 +1,366 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    http://docs.cilium.io
+
+.. _ipam_eni:
+
+#######
+AWS ENI
+#######
+
+The AWS ENI allocator is specific to Cilium deployments running in the AWS
+cloud and performs IP allocation based on IPs of `AWS Elastic Network Interfaces (ENI)
+<https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html>`__ by
+communicating with the AWS EC2 API.
+
+The architecture ensures that only a single operator communicates with the EC2
+service API to avoid rate-limiting issues in bigger clusters. A pre-allocation
+watermark allows to have IP addresses available for use on nodes at all time
+without requiring to contact the EC2 API when a new pod is scheduled in the
+cluster.
+
+************
+Architecture
+************
+
+.. image:: eni_arch.png
+    :align: center
+
+The AWS ENI allocator builds on top of the CRD-backed allocator. Each node
+creates a ``ciliumnodes.cilium.io`` custom resource matching the node name when
+Cilium starts up for the first time on that node. It contacts the EC2 metadata
+API to retrieve instance ID, instance type, and VPC information and populates
+the custom resource with this information. ENI allocation parameters are
+provided as agent configuration option and are passed into the custom resource
+as well.
+
+The Cilium operator listens for new ``ciliumnodes.cilium.io`` custom resources
+and starts managing the IPAM aspect automatically. It scans the EC2 instances
+for existing ENIs with associated IPs and makes them available via the
+``spec.ipam.available`` field. It will then constantly monitor the used IP
+addresses in the ``status.ipam.used`` field and automatically create ENIs and
+allocate more IPs as needed to meet the IP pre-allocation watermark. This ensures
+that there are always IPs available
+
+The selection of subnets to use for allocation as well as attachment of
+security groups to new ENIs can be controlled separately for each node. This
+makes it possible to hand out pod IPs with differing security groups on
+individual nodes.
+
+The corresponding datapath is described in section :ref:`datapath_eni`.
+
+*************
+Configuration
+*************
+
+* The Cilium agent and operator must be run with the option ``--ipam=eni`` or
+  the option ``ipam: eni``  must be set in the ConfigMap. This will enable ENI
+  allocation in both the node agent and operator.
+
+* In most scenarios, it makes sense to automatically create the
+  ``ciliumnodes.cilium.io`` custom resource when the agent starts up on a node
+  for the first time. To enable this, specify the option
+  ``--auto-create-ciliumnode-resource`` or  set
+  ``auto-create-ciliumnode-resource: "true"`` in the ConfigMap.
+
+* It is generally a good idea to enable metrics in the Operator as well with
```

Comment:
Perhaps note that Prometheus should be configured as well and link to that documentation here.